src/App.tsx

```diff
@@ -1,24 +1,284 @@
+  updatedField: keyof Todo,
+) {
+  return todos.map(todo => {
+    if (todo.id === updatedTodo.id) {
+      return {
+        ...todo,
+        [updatedField]: updatedTodo[updatedField],
+      };
+    }
+
+    return todo;
+  });
+}
 
 export const App: React.FC = () => {
+  const { todos, setTodos } = useContext(TodosContext);
+  const {
+    todosIdToDelete,
+    setTodosIdToDelete,
+    setTodosIdToUpdate,
+  } = useContext(LoadingTodosContext);
+
+  const [isLoadingTodos, setIsLoadingTodos] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+
+  const [isLoadingAddTodo, setIsLoadingAddTodo] = useState(false);
+
+  const formInput = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (!todosIdToDelete.length && !isLoadingAddTodo && formInput.current) {
+      formInput.current.focus();
+    }
+  }, [isLoadingAddTodo, todosIdToDelete]);
+
+  const handleChangeNewTodoTitle = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setNewTodoTitle(event.target.value);
+  };
+
+  const hasCompletedTodo = useMemo(() => {
+    return todos.some(todo => todo.completed);
+  }, [todos]);
+
+  const notCompletedTodosLength = useMemo(() => {
+    return todos.filter(todo => !todo.completed).length;
+  }, [todos]);
+
+  const isCompletedAll = useMemo(() => {
+    return todos.every(todo => todo.completed);
+  }, [todos]);
+
+  const handelSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setErrorMessage('');
+    setIsLoadingAddTodo(true);
+
+    const newTodo = {
+      userId: USER_ID,
+      title: newTodoTitle.trim(),
+      completed: false,
+    };
+
+    setTempTodo({
+      ...newTodo,
+      id: 0,
+    });
+
+    addNewTodo(newTodo)
+      .then(response => {
+        setNewTodoTitle('');
+
+        return response;
+      })
+      .then(currentTodo => {
+        setTodos((prevState) => [...prevState, currentTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsLoadingAddTodo(false);
+        setTempTodo(null);
+      });
+  };
+
+  useEffect(() => {
+    setIsLoadingTodos(true);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      })
+      .finally(() => {
+        setIsLoadingTodos(false);
+      });
+  }, []);
+
+  const handelDeleteTodoById = (todoId: number) => {
+    setTodosIdToDelete(prevState => [...prevState, todoId]);
+
+    return deleteTodoById(todoId)
+      .then(() => {
+        setTodos(prevState => prevState.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+        throw new Error();
```

Comment:
It is also worth passing the error text to the `Error` constructor, because without it there is no point in throwing out the error (below is a screenshot of the error from your application):
<img width="425" alt="image" src="https://github.com/mate-academy/react_todo-app-with-api/assets/54860641/bdc2e36c-6733-4e65-ab3b-95c9f5be24fa">
Fix this in all code.